### PR TITLE
ddl: Migrate tests of TiFlash Cluster Manager into testify

### LINKED ISSUE #31844: migrate test-infra to testify for `ddl/ddl_tiflash_test.go`
## Enhancement

Sub task of #27180

### PR DESCRIPTION
**Pull request by CalvinNeo**
### What problem does this PR solve?

Issue Number: close #31844

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: ddl/ddl_tiflash_test.go [hunk 1]
```diff
@@ -81,31 +80,35 @@ func (s *tiflashDDLTestSuite) SetUpSuite(c *C) {
...
 	s.tiflash.Unlock()
 	s.dom.Close()
-	err := s.store.Close()
-	c.Assert(err, IsNil)
+	s.store.Close()
 	ddl.PollTiFlashInterval = 2 * time.Second
 }
 
-func ChangeGCSafePoint(tk *testkit.TestKit, t time.Time, enable string, lifeTime string) {
+func ChangeGCSafePoint(tk *testkit2.TestKit, t time.Time, enable string, lifeTime string) {
```
Comment (hawkingrei) [inline ddl/ddl_tiflash_test.go:111]: Should be `testkit`

File: ddl/ddl_tiflash_test.go [hunk 2]
```diff
@@ -142,21 +144,21 @@ func (s *tiflashDDLTestSuite) CheckFlashback(tk *testkit.TestKit, c *C) {
...
-	c.Assert(err, IsNil)
-	c.Assert(tb, NotNil)
+	require.NoError(t, err)
+	require.NotNil(t, tb)
 	if tb.Meta().Partition != nil {
 		for _, e := range tb.Meta().Partition.Definitions {
 			ruleName := fmt.Sprintf("table-%v-r", e.ID)
 			_, ok := s.tiflash.GetPlacementRule(ruleName)
-			c.Assert(ok, Equals, true)
+			require.Equal(t, true, ok)
```
Comment (hawkingrei) [inline ddl/ddl_tiflash_test.go:156]: require.True

Ditto other.

**Comment by tisonkun** [comment]
This PR doesn't need a Release Note I think.

File: ddl/ddl_tiflash_test.go [hunk 3]
```diff
@@ -38,7 +38,8 @@ import (
 	"github.com/pingcap/tidb/store/gcworker"
 	"github.com/pingcap/tidb/store/mockstore"
 	"github.com/pingcap/tidb/store/mockstore/unistore"
-	"github.com/pingcap/tidb/util/testkit"
+	testkit "github.com/pingcap/tidb/testkit"
```
Comment (hawkingrei) [inline ddl/ddl_tiflash_test.go:41]:
```
"github.com/pingcap/tidb/testkit"
```

Comment (tisonkun) [inline ddl/ddl_tiflash_test.go:41]:
```suggestion
	"github.com/pingcap/tidb/testkit"
```

@CalvinNeo you don't resolve the problem actually :P

File: ddl/ddl_tiflash_test.go [hunk 4]
```diff
@@ -50,14 +51,12 @@ type tiflashDDLTestSuite struct {
...
 
-var _ = SerialSuites(&tiflashDDLTestSuite{})
-
 const (
 	RoundToBeAvailable               = 2
 	RoundToBeAvailablePartitionTable = 3
 )
 
-func (s *tiflashDDLTestSuite) SetUpSuite(c *C) {
+func SetUpHelper(s *tiflashDDLTestSuite) error {
```
Comment (tisonkun) [inline ddl/ddl_tiflash_test.go:59]: We don't write test in this style. Please checkout other test file in testify.

File: ddl/ddl_tiflash_test.go (hunk 3 above)
Comment (CalvinNeo) [reply ddl/ddl_tiflash_test.go]: Have not pushed latest commits...sorry

File: ddl/ddl_tiflash_test.go (hunk 4 above)
Comment (CalvinNeo) [reply ddl/ddl_tiflash_test.go:59]: I don't quite follow... I am trying to follow examples like `TestAnalyzeIndexExtractTopN` before migrate the test to testify. `tiflashDDLTestSuite` is something from gocheck, it only act as a combination of context now.

File: ddl/ddl_tiflash_test.go [hunk 5]
```diff
@@ -81,27 +86,31 @@ func (s *tiflashDDLTestSuite) SetUpSuite(c *C) {
...
 
-func (s *tiflashDDLTestSuite) TearDownSuite(c *C) {
+func teardownHelper(s *tiflashContext) {
 	s.tiflash.Lock()
 	s.tiflash.StatusServer.Close()
 	s.tiflash.Unlock()
 	s.dom.Close()
-	err := s.store.Close()
-	c.Assert(err, IsNil)
+	s.store.Close()
```
Comment (xhebox) [inline ddl/ddl_tiflash_test.go:113]: The assertion of nil should be kept.

File: ddl/ddl_tiflash_test.go [hunk 6]
```diff
@@ -125,13 +134,12 @@ func ChangeGCSafePoint(tk *testkit.TestKit, t time.Time, enable string, lifeTime
...
 }
 
-func (s *tiflashDDLTestSuite) CheckPlacementRule(rule placement.TiFlashRule) bool {
-	return s.tiflash.CheckPlacementRule(rule)
+func CheckPlacementRule(tiflash *infosync.MockTiFlash, rule placement.TiFlashRule) bool {
+	return tiflash.CheckPlacementRule(rule)
 }
 
-func (s *tiflashDDLTestSuite) CheckFlashback(tk *testkit.TestKit, c *C) {
+func CheckFlashback(s *tiflashContext, tk *testkit.TestKit, t *testing.T) {
```
Comment (xhebox) [inline ddl/ddl_tiflash_test.go:141]: Could be changed to a method of `tiflashContext`

File: ddl/ddl_tiflash_test.go [hunk 7]
```diff
@@ -81,27 +86,31 @@ func (s *tiflashDDLTestSuite) SetUpSuite(c *C) {
...
+	}
 	s.dom.SetStatsUpdating(true)
 
 	log.Info("Mock stat", zap.Any("infosyncer", s.dom.InfoSyncer()))
+
+	return nil
 }
 
-func (s *tiflashDDLTestSuite) TearDownSuite(c *C) {
+func teardownHelper(s *tiflashContext) {
```
Comment (xhebox) [inline ddl/ddl_tiflash_test.go:108]: Could be a method of `tiflashContext`

File: ddl/ddl_tiflash_test.go [hunk 8]
```diff
@@ -173,17 +181,21 @@ func TempDisableEmulatorGC() func() {
...
 	return func() {
 		ddl.PullTiFlashPdTick.Store(originValue)
 	}
 }
 
 // Run all kinds of DDLs, and will create no redundant pd rules for TiFlash.
-func (s *tiflashDDLTestSuite) TestTiFlashNoRedundantPDRules(c *C) {
+func TestTiFlashNoRedundantPDRules(t *testing.T) {
+	s := createTiFlashContext(t)
+	defer teardownHelper(s)
```
Comment (xhebox) [inline ddl/ddl_tiflash_test.go:194]: I think you could return `tiflashcontext, testkit, cleanfn := createTiFlashContext(t); defer cleanfn()`. Some other testify migration are doing so.

**Review by xhebox — COMMENTED** [review]
LGTM otherwise

File: ddl/ddl_tiflash_test.go (hunk 7 above)
Comment (CalvinNeo) [reply ddl/ddl_tiflash_test.go:108]: Will migrate all `Change*` funcs

**Review by xhebox — APPROVED** [review]

**Review by hawkingrei — APPROVED** [review]

**Comment by hawkingrei** [comment]
@tisonkun  PTAL

**Review by tisonkun — APPROVED** [review]

**Pull request merged**